src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -618,10 +624,10 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
                             Log.RequestBodyIOException(httpContext, ex);
                             return;
                         }
-                        catch (InvalidDataException ex)
+                        catch (Exception ex) when (ex is InvalidDataException || ex is JsonException)
```

Comment:
From looking at [the tests here](https://github.com/dotnet/aspnetcore/blob/f6efa13e71d3f82f149d9af8f7c47860d87ffa9c/src/Http/Http.Extensions/test/HttpRequestJsonExtensionsTests.cs) it looks like the deserialization process will throw one of three exceptions:

- InvalidOperationException
- JsonException
- TaskCancelledException

And [the docs for the JsonSerializer](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserializeasync?view=net-5.0) indicate that it can also throw:

- NotSupportedException
- ArgumentNullException

I think for this case we probably want to catch the NotSupportedException and the JsonException since both indicate an inability to parse the response.

Reply:
Not sure about `NotSupportedException`. That one feels like it's an issue on the server side. You could imagine adding a converter, not changing the client's payload and things beginning to work all of a sudden.

Reply:
I think that the `JsonException` is the only thing listed that clearly points to the input being the error. All the rest looks like API misusage on the server or server-side cancellation.

Reply:
Fair enough. I agree with the sentiment that we can safely remove the check for `InvalidDataException` and just check for `JsonException`.

Reply:
@captainsafia @Tratcher I've updated this to only catch JsonExceptions now.